### PR TITLE
Refactor: Discord events as modules, role provider modules, commands receive Eris Messages

### PR DESCRIPTION
**Pull request by kevinlul**
We adopt a per-file approach to events. Unfortunately this means that we'll have to add the typings and make sure they match Eris. Currently passing in dependencies via a higher-order function as a sort of dependency injection but this is very much changeable as we figure out the best approach.

All custom typings that messageCreate uses have been pulled over in an effort to isolate; it remains to be seen whether copying them will be fruitful.

Role providers are to be extracted based on previous work in the 46 branch. OrganiserRoleProvider is pretty much based on that, and it allows us to rip all that functionality out and remove a lot of indirection that made finding these relationships in Code very difficult. This forces me to temporarily add the original message to DiscordMessageIn until that's sorted out.  (Bug with participant role takeover is a single-line deletion here!)

We'll also create a ParticipantRoleProvider that maps tournament id -> role id -> server id and go from there. As a side effect of having to insert the original Message back in, it might be destroyed in this PR or an immediate follow-up. I think the reaction handlers are _probably_ out of scope, as well as anything to do with removing `config.json`. 

Beware CI fails!

## Checklist

- [x] I am following the [contributing guidelines]( https://github.com/AlphaKretin/emcee-tournament-bot/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AlphaKretin** [comment]
I just caught up on this so far, I definitely approve of how this is developing. I agree reaction handlers are out of scope since the event callbacks are set up at arbitrary times instead of at launch.

File: src/role/participant.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	 */
+	protected async lazyGet(tournament: Tournament): Promise<ParticipantRole> {
+		if (tournament.id in this.roleCache) {
+			return this.roleCache[tournament.id];
+		}
+		const server = this.bot.guilds.get(tournament.server);
+		if (!server) {
+			// TODO: determine what bizarre scenario would result in this
+			throw new MiscInternalError(
+				`Could not find server ${tournament.server} as registered with Tournament ${tournament.server}.`
```
Comment (AlphaKretin) [inline src/role/participant.ts:62]: i think the second parameter here should be `tournament.name`, not `tournament.server`

File: test/commandHandler.ts [hunk 2]
```diff
@@ -12,39 +12,7 @@ const mockTournament = new TournamentMock();
...
-		);
-	});
-	it("Argument validation", async function () {
-		await discord.simMessage("mc!create", "error");
-		const response = discord.getResponse("error");
-		expect(response).to.equal("Usage: create name|description");
-	});
-});
 describe("Tournament creation commands", function () {
-	it("Create tournament", async function () {
```
Comment (AlphaKretin) [inline test/commandHandler.ts:37]: i assume the tests deleted here are to be restored as proper unit test files before this PR is finished?

**Review by AlphaKretin — COMMENTED** [review]
So I understand removing the Discord indirection because it just impedes proper testing to enable bad testing, and adds more layers for things to get mixed up, but it does leave us reliant on Eris' specific interface and if it were to change we'd have to change our code everywhere and not just the wrapper.

I'm not sure how I feel about the amount of logic moved into commands for the CSV generators, but I do understand the desire to move them out of TournamentManager so it's not like there's anywhere better for them.

Overall happy with the progress on this so far

File: test/commandHandler.ts (hunk 2 above)
Comment (kevinlul) [reply test/commandHandler.ts:37]: They already did

**Comment by kevinlul** [comment]
Having to change code if an upstream breaks compatibility is pretty much the baseline cost of using third-party dependencies. Has this specifically happened with Eris before and what was the extent? Is it worse than having to create the same abstraction in three places? One of the indirections simply changed the names of fields within a TypeScript type (i.e. almost zero runtime impact besides remapping properties), which seems an unnecessary exercise unless an adapter pattern was truly required to work with a different third-party dependency.

I think it's appropriate that the logic for the commands stays in the commands, because then you can find out what a command does by reading the source code for a command, and you can meaningfully unit test a command instead of testing whether it calls an implementation, which is not as meaningful. The next steps in the direction of this PR (but not in this PR) would get rid of both the wrapper and the tournament manager since they were basically just god objects.

**Comment by kevinlul** [comment]
We can do the unit tests here to replace everything in `tests/commandHandler` or we can do it in later PRs.

**Comment by AlphaKretin** [comment]
> Having to change code if...
> 
> I think it's appropriate that...

Both of these points make sense, I think I just had bad preconceived notions about what "good design" looks like. Carry on.

File: src/role/participant.ts (hunk 1 above)
Comment (AlphaKretin) [reply src/role/participant.ts:62]: This issue is extant.

**Review by AlphaKretin — CHANGES_REQUESTED** [review]
One minor issue outstanding, but otherwise, as discussed through reviews while the PR was in progress, I'm happy with where this is at. Replacing the tests in a separate PR sounds good to me.

**Review by AlphaKretin — APPROVED** [review]

**Pull request merged**